### PR TITLE
Reporting pipeline execution status through an e-mail.

### PR DESCRIPTION
**Pull request by amwolff**

### CODE REVIEW

File: mail/mail.go [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+import (
+	"bytes"
+	"html/template"
+
+	"github.com/datainq/jobhub"
+	"github.com/sirupsen/logrus"
+	"gopkg.in/gomail.v2"
+)
+
+// type SMS struct {
```
Comment (orian) [inline mail/mail.go]: remove comment

File: mail/mail.go [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+	Subject      string
+	Host         string
+	Port         int
+	Username     string
+	Password     string
+	TemplatePath string
+	Log          logrus.FieldLogger
+}
+
+func parseTemplate(templatePath string, data interface{}) (string, error) {
```
Comment (orian) [inline mail/mail.go]: This function does both: parsing and executing template. The name does not correspond to that.

File: mail/mail.go [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+	t := template.Must(template.ParseFiles(templatePath))
+	buf := new(bytes.Buffer)
+	if err := t.Execute(buf, data); err != nil {
+		return "", err
+	}
+	return buf.String(), nil
+}
+
+func (m *Mail) SendStatus(pipelineStatus jobhub.PipelineStatus) {
+	if len(m.Subject) <= 0 {
```
Comment (orian) [inline mail/mail.go]: len cannot be < 0

File: mail/mail.go [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+func parseTemplate(templatePath string, data interface{}) (string, error) {
+	t := template.Must(template.ParseFiles(templatePath))
+	buf := new(bytes.Buffer)
+	if err := t.Execute(buf, data); err != nil {
+		return "", err
+	}
+	return buf.String(), nil
+}
+
+func (m *Mail) SendStatus(pipelineStatus jobhub.PipelineStatus) {
```
Comment (orian) [inline mail/mail.go]: This function does a lot of things. It parses, loads file, renders template and sends email. 
Let's replace `Subject` and `TemplatePath` fields by `*template.Template` and assume someone set if from outside.

```
   SubjectTmpl *template.Template
   BodyTmpl *template.Template
```

Then you can separate rendering templates from sending an email.

Also consider, that right now the `Mail` is a mix of email to be send (Subject field) and a config to be used. They are two different things, worth splitting.

File: mail/mail_test.go [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+	"testing"
+)
+
+func TestParseTemplate(t *testing.T) {
+	defer func() {
+		if r := recover(); r == nil {
+			t.Errorf("The code did not panic.")
+		}
+	}()
+	parseTemplate("./mail_template.html", nil)
```
Comment (orian) [inline mail/mail_test.go]: I think, this test always succeeds, even if you remove `./mail_template.html` ;-)

**Review by orian — COMMENTED** [review]
PTAL

File: mail/mail.go (hunk 3 above)
Comment (amwolff) [reply mail/mail.go]: Done.

File: mail/mail.go (hunk 1 above)
Comment (amwolff) [reply mail/mail.go]: Done.

File: mail/mail.go (hunk 2 above)
Comment (amwolff) [reply mail/mail.go]: Done.

File: mail/mail.go (hunk 4 above)
Comment (amwolff) [reply mail/mail.go]: Done.

File: mail/mail.go [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+	"github.com/datainq/jobhub"
+	"github.com/sirupsen/logrus"
+	"gopkg.in/gomail.v2"
+)
+
+type Mail struct {
+	From string
+	To   string
+
+	Host     string
```
Comment (orian) [inline mail/mail.go]: Please move subject and body above the email config

**Review by orian — CHANGES_REQUESTED** [review]

File: mail/mail.go (hunk 6 above)
Comment (amwolff) [reply mail/mail.go]: Done.

**Review by orian — APPROVED** [review]
LGTM

**Pull request merged**